yesod-core/src/Yesod/Core/Internal/Run.hs

```diff
@@ -116,7 +117,7 @@ basicRunHandler rhe handler yreq resState = do
             return (HCContent defaultStatus tc))
         (\e ->
             case fromException e of
-                Just e' -> return e'
+                Just (AnnotatedException _ e') -> return e'
                 Nothing -> HCError <$> toErrorHandler e)
```

Comment:
OK, digging into this some more:

```haskell
-- | Convert a synchronous exception into an ErrorResponse
toErrorHandler :: SomeException -> IO ErrorResponse
toErrorHandler e0 = handleAny errFromShow $
    case fromException e0 of
        Just (HCError x) -> evaluate $!! x
        _ -> errFromShow e0
```

So I think *this* function needs to also check for `AnnotatedException` - since the `fromException e0 :: Maybe HCContents` won't work if the underlying application code is calling `throwWithCallStack`.

```haskell
-- | Generate an @ErrorResponse@ based on the shown version of the exception
errFromShow :: SomeException -> IO ErrorResponse
errFromShow x = do
  text <- evaluate (T.pack $ show x) `catchAny` \_ ->
          return (T.pack "Yesod.Core.Internal.Run.errFromShow: show of an exception threw an exception")
  return $ InternalError text
```

Fortunately this should work fine - perhaps we want to unwrap the `AnnotatedException` before showing it?